### PR TITLE
[TM] add spec for PlatformConstants

### PR DESCRIPTION
**Pull request by sattaman**
## Summary

part of #24875.

## Changelog

[General] [Added] - add TM spec for PlatformConstants

## Test Plan

Tests + flow pass

### CODE REVIEW

File: Libraries/Utilities/Platform.android.js [hunk 1]
```diff
@@ -20,19 +19,17 @@ export type PlatformSelectSpec<A, D> = {
 const Platform = {
   OS: 'android',
   get Version() {
-    const constants = NativeModules.PlatformConstants;
-    return constants && constants.Version;
+    const { Version } = NativePlatformConstants.getConstants();
+    return Version || 0;
```
Comment (sattaman) [inline Libraries/Utilities/Platform.android.js]: I wasn't sure if there was a better was of forcing the value going from here to be non nullable

**Comment by sattaman** [comment]
I didn't realise when picking this up @jarvisluong It looks like it might it might have been better if I'd picked up the ios part of PlatformConstants too since its all the same module .

**Comment by jarvisluong** [comment]
@sattaman right, that is fine too!

File: Libraries/Core/ReactNativeVersionCheck.js [hunk 2]
```diff
@@ -46,6 +47,8 @@ exports.checkVersions = function checkVersions(): void {
 function _formatVersion(version): string {
   return (
     `${version.major}.${version.minor}.${version.patch}` +
-    (version.prerelease !== null ? `-${version.prerelease}` : '')
+    (version.prerelease !== null && version.prerelease !== undefined
```
Comment (uqmessias) [inline Libraries/Core/ReactNativeVersionCheck.js]: In order to keep it less verbose, what do you think of checking for `version.prerelease != undefined` instead? By doing that, we can check if it's different from `null` and `undefined`

Comment (sattaman) [reply Libraries/Core/ReactNativeVersionCheck.js]: Sure, I guess `version.prerelease &&` will serve the same purpose.. I wasn't sure if this would have be frowned upon

Comment (uqmessias) [reply Libraries/Core/ReactNativeVersionCheck.js]: Unfortunately no.
`version.prerelease` will be very different from simply `version.prerelease != undefined` because:

```javascript
const cond1 = version.prerelease;
const cond2 = !!version.prerelease;
const cond3 = (version.prerelease !== 0 && version.prerelease !== undefined && version.prerelease !== null && version.prerelease !== '' && version.prerelease !== false);
// cond1 === cond2 === cond3

const cond4 = version.prerelease != undefined;
const cond5 = version.prerelease != null;
const cond6 = version.prerelease !== undefined && version.prerelease !== null;
// cond4 === cond5 === cond6
```

Comment (uqmessias) [reply Libraries/Core/ReactNativeVersionCheck.js]: This is a minor issue. In the end, it will be a simplification of `version.prerelease !== null && version.prerelease !== undefined`

File: RNTester/js/TouchableExample.js [hunk 3]
```diff
@@ -18,17 +18,15 @@ const {
   Text,
   TouchableHighlight,
   TouchableOpacity,
-  NativeModules,
+  NativePlatformConstants,
```
[the remark this answers is not in the document]
Comment (sattaman) [reply RNTester/js/TouchableExample.js]: I'm wondering if I'm doing the right thing updating this here?

File: Libraries/Core/ReactNativeVersionCheck.js (hunk 2 above)
Comment (sattaman) [reply Libraries/Core/ReactNativeVersionCheck.js]: I've updated, many thanks for taking the time to explain. I'm finding I've got 152 tests failing locally, and I'm not sure why, also flow isn't seeming to like what I've done. Is there anything obvious wrong with what I've done?

File: RNTester/js/TouchableExample.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sattaman) [reply RNTester/js/TouchableExample.js]: I've reverted this as presuming not

File: yarn.lock [hunk 4]
```diff
@@ -1163,6 +1163,7 @@
 "@reactions/component@^2.0.2":
   version "2.0.2"
   resolved "https://registry.yarnpkg.com/@reactions/component/-/component-2.0.2.tgz#40f8c1c2c37baabe57a0c944edb9310dc1ec6642"
+  integrity sha1-QPjBwsN7qr5XoMlE7bkxDcHsZkI=
```
Comment (fkgozali) [inline yarn.lock]: Can you undo this one? seems unrelated to the change

File: Libraries/Utilities/NativePlatformConstants.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import type {TurboModule} from 'RCTExport';
+import * as TurboModuleRegistry from 'TurboModuleRegistry';
+import Platform from 'Platform';
+
+export interface Spec extends TurboModule {
+  // Exported methods.
+  +getConstants: () => {|
+    // Common interface
+    isTesting: boolean,
+    reactNativeVersion: {
```
Comment (fkgozali) [inline Libraries/Utilities/NativePlatformConstants.js]: let this be strict object as well:

```
reactNativeVersion: {|
  ...
|},
```

File: Libraries/Utilities/NativePlatformConstants.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    // iOS only
+    forceTouchAvailable?: boolean,
+    osVersion?: number,
+    systemName?: string,
+    interfaceIdiom?: 'phone' | 'pad' | 'tv' | 'carplay' | 'unknown',
+  |};
+
+  // Android only
+  +getAndroidID?: () => string;
```
Comment (fkgozali) [inline Libraries/Utilities/NativePlatformConstants.js]: make this required (no ?)

File: Libraries/Utilities/NativePlatformConstants.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    },
+
+    // Android only
+    Version?: number,
+    Release?: string,
+    Serial?: string,
+    Fingerprint?: string,
+    Model?: string,
+    ServerHost?: string,
+    uiMode?: 'tv' | 'car' | 'desk' | 'watch' | 'normal' | 'unknown',
```
Comment (RSNara) [inline Libraries/Utilities/NativePlatformConstants.js]: Our Codegen doesn't support enums. Can you change this to `string`?

File: Libraries/Utilities/NativePlatformConstants.js (hunk 6 above)
Comment (RSNara) [reply Libraries/Utilities/NativePlatformConstants.js]: Shouldn't this be optional because iOS doesn't expose this method?

Comment (fkgozali) [reply Libraries/Utilities/NativePlatformConstants.js]: Optional or not, codegen still generates the method, so the function existence check is now meaningless. Might as well make it required.

**Comment by sattaman** [comment]
@fkgozali I've updated as requested
